Consistently initialize `Exception.name` across back-ends

zerbina opened the pull request:
Creating an `Exception` without raising it left it's `name` field as nil
on the JS and VM back-ends. For the C targets, the name initialization
was injected during code-gen.

Initializing the `name` field now either happens in `newException` or
when raising the exception and thus works the same across all back-ends.
Creating an exception without `newException` doesn't set the `name`
field on the C back-ends anymore.

In addition, when raising an exception on the JS target, the exception's
name is not overridden if it's non-nil.

This is a breaking change.

---
<!-- Note: section break (`---`) onwards is not in CI merge commit -->

If the exception's name should also be automatically set when not using `newException`, I'd propose doing this in `transf` instead of during code-gen

saem commented:
This feels more consistent, if some platform compatibility bits come up we can consider those as they arise and revisit.

Bors r+

zerbina commented:
The failing cpp test is due to the recent `sem` [change](https://github.com/nim-works/nimskull/pull/290)

saem commented:
> The failing cpp test is due to the recent `sem` [change](https://github.com/nim-works/nimskull/pull/290)

Exclude CPP from the test and leave a comment in the test with a comment (`# knownIssue: ...`).

zerbina commented:
The failing cpp test should be fixed by [https://github.com/nim-works/nimskull/pull/294](https://github.com/nim-works/nimskull/pull/294). I'll rebase, once it's merged

saem commented:
Merge in progress. 🎉

saem commented:
bors r+

zerbina committed:
exceptions: consistently initialize `name` field across back-ends

Creating an `Exception` without raising it left it's `name` field as nil
on the JS and VM back-ends. For the C targets, the name initialization
was injected during code-gen.

Initializing the `name` field now either happens in `newException` or
when raising the exception and thus works the same across all back-ends.
Creating an exception without `newException` doesn't set the `name`
field on the C back-ends anymore.

In addition, when raising an exception on the JS target, the exception's
name is not overriden if it's non-nil.

This is a breaking change.
